gc/api/CMakeLists.txt

```diff
@@ -0,0 +1,52 @@
+include(OmrAssert)
+
+omr_assert(TEST OMR_GC_EXTENDED_API)
+
+###
+### OMR GC
+###
+
+target_sources(omrgc
+	PRIVATE
+		${CMAKE_CURRENT_SOURCE_DIR}/src/OMR_GC_System.cpp
+		${CMAKE_CURRENT_SOURCE_DIR}/glue/MarkingDelegate.cpp
+		${CMAKE_CURRENT_SOURCE_DIR}/glue/EnvironmentDelegate.cpp
+)
+
+target_include_directories(${OMR_GC_GLUE_TARGET} BEFORE
+	INTERFACE
+		${CMAKE_CURRENT_SOURCE_DIR}/glue/
+		${CMAKE_CURRENT_SOURCE_DIR}/include/
+)
+
+target_include_directories(omrgc BEFORE
+	PUBLIC
+		TEST_OMRGC
+		glue/
+		include/
+)
+
+# https://cmake.org/cmake/help/v3.6/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES
+target_compile_features(omrgc
+	PUBLIC
+		cxx_constexpr
+		# cxx_auto_type
+		cxx_deleted_functions
+		cxx_delegating_constructors
```

Comment:
Are all of these available on the OMR minimum supported compiler levels?

Reply:
Most of these features are actually for Om, not the object scanner / context. I can probably get rid of most (if not all) for this PR.

To answer the question though, I believe most of these features are _not_ supported, so that will be a problem for the Om PR.

Reply:
Well after looking over the code, this is what I'm seeing:
* noexcept: in my case, I can just use the deprecated `throw()` syntax, or just get rid of the exception specifiers entirely, or define a OMR_NOEXCEPT symbol depending on the compiler support.
* nullptr: can be replaced by `NULL`, except in the `OMR::GC::Ref<T>`, where it is used in the interface. I can remove Ref from this PR, and deal with it later.
* delegating constructors: I can rewrite some code to not use this feature, it shouldn't be difficult.
* constexpr: this one is probably going to hurt to remove, I'm using constexpr a lot. APIs in Om will have to drop constexpr too. I think it's worthwhile to find a way to support constexpr when available. Maybe we can test for `__cplusplus` to define`OMR_CXX11_CONSTEXPR`, and fall back to `inline` if its not available in the current compilation unit? That might break ODR though.
* Static assert / type traits: they're such a good safety net that I'd be sad to remove them. Maybe I can hide the static asserts under an ifdef like `OMR_HAVE_CXX11_STATIC_ASSERT`, or define an `OMR_STATIC_ASSERT` macro. It's harder to replace the type traits though. I'm using `std::is_standard_layout`.
* range-based for loop: I'm using it to iterate the marking functions attached to the context, but I can get rid of this no-problem.

Reply:
There is a module in cmake for automatically defining portable wrappers around c++ features, called [WriteCompilerDetectionHeader](https://cmake.org/cmake/help/v3.12/module/WriteCompilerDetectionHeader.html#module:WriteCompilerDetectionHeader).

It looks extremely useful, but it would (using a symbol prefix of OMR) define symbols such as `OMR_COMPILER_IS_SCO`, which might conflict with the compiler project. Alternatively, using a prefix of `OMR_CXX` would define symbols like `OMR_CXX_COMPILER_CXX_CONSTEXPR` which is not great either.

Something like this though, would be a good way to conditionally support newer cxx features when available.

Reply:
I opened an issue for implementing portable c++11 feature wrappers:
https://github.com/eclipse/omr/issues/2955

Reply:
There is some new discussion ongoing about how (or if) we'll be making use of C++11 in the future. Until we land on a decision, I'd like to leave the C++11 isms in this PR as-is. Since this code is disabled by default, and explicitly opt-in only, this should have no impact on existing consumers of OMR. When we figure out what to do about C++11, I'll update this code to follow our new coding practices.